service/dap/server.go

```diff
@@ -2957,6 +3037,54 @@ func (s *Server) doRunCommand(command string, asyncSetupDone chan struct{}) {
 	}
 }
 
+func (s *Server) run(command string, asyncSetupDone chan struct{}) (*api.DebuggerState, error) {
+	// TODO(polina): it appears that debugger.Command doesn't always close
+	// asyncSetupDone (e.g. when having an error next while nexting).
+	// So we should always close it ourselves just in case.
+	defer s.asyncCommandDone(asyncSetupDone)
+	s.setRunning(true)
+	defer s.setRunning(false)
+
+	runningStep := command != api.Continue
+	state, _ := s.debugger.State(true)
+	if state != nil {
+		// If state.NextInProgress, then we are in the middle of a step,
+		// regardless of the command.
+		runningStep = runningStep || state.NextInProgress
+	}
+	// The handleRequest loop will release resumeMu after asyncSetupDone is closed.
+	state, err := s.debugger.Command(&api.DebuggerCommand{Name: command}, asyncSetupDone)
+	for state != nil && state.CurrentThread != nil && err == nil {
+		if bp := state.CurrentThread.Breakpoint; bp != nil && bp.Tracepoint {
+			// If this is a log point, we want to log the message and continue execution.
+			s.handleLogPoint(bp.ID)
+			// Only resume execution if continue was the command used to run the program.
+			// Otherwise, the program will go past the step, next, stepout requests.
+			// TODO(suzmue): have s.debugger.Command() not clear internal breakpoints when
+			// hitting another breakpoint.
+			if !runningStep || state.NextInProgress {
+				state, err = s.resume()
```

Comment:
It bothers me that the first resume is done differently than every resume thereafter.

Reply:
This is necessary because for the first continue command, the handleRequest loop is waiting for the continue command to start.

All subsequent resumes could be happening while the handleRequest loop is doing something else .

Reply:
So is the only difference that the first time you don't acquire resumeMu? What if you acquired it anyway?

Reply:
We don't acquire resumeMu because handleRequest is already holding it (this could be changed by moving where the lock is acquired in handleRequest) and we close asyncSetupDone. handleRequest is waiting for asyncSetupDone to be closed before it checks for more requests. 

I could refactor this to potentially use the same resume() function, but there would still be a difference of which channel gets closed.

Reply:
The other thing that is different is the first resume could be api.Next, api.Continue, api.StepIn, etc, whereas all subsequent ones are api.Continue

Reply:
I refactored this to always use resume, and moved the locking to make this possible.